src/poetry/vcs/git/backend.py

```diff
@@ -313,21 +314,47 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
             if isinstance(e, AssertionError) and "Invalid object name" not in str(e):
                 raise
 
-            logger.debug(
-                "\nRequested ref (<c2>%s</c2>) was not fetched to local copy and"
-                " cannot be used. The following error was"
-                " raised:\n\n\t<warning>%s</>",
-                refspec.key,
-                e,
-            )
+            short_ref_found = False
 
-            raise PoetryConsoleError(
-                f"Failed to clone {url} at '{refspec.key}', verify ref exists on"
-                " remote."
-            )
+            if refspec.is_sha_short:
+                commit = cls._find_object_by_sha_prefix(local, refspec.key.encode())
+                if commit is not None:
+                    logger.debug(
+                        "\nResolved short SHA <c2>%s</c2> as <c2>%s</c2>",
+                        refspec.key,
+                        commit.id,
+                    )
+                    local.refs[b"HEAD"] = commit.id
+                    try:
+                        with local:
+                            local.reset_index()
+                        short_ref_found = True
+                    except (AssertionError, KeyError) as e:
+                        logger.debug(
+                            "\nRequested ref (<c2>%s</c2>) was not fetched to local copy and"
+                            " cannot be used. The following error was"
+                            " raised:\n\n\t<warning>%s</>",
+                            refspec.key,
+                            e,
+                        )
+
+            if not short_ref_found:
+                raise PoetryConsoleError(
+                    f"Failed to clone {url} at '{refspec.key}', verify ref exists on"
+                    " remote."
+                )
 
         return local
 
+    @classmethod
+    def _find_object_by_sha_prefix(cls, repo: Repo, sha_prefix: bytes) -> Commit | None:
+        for sha in repo.object_store:
```

Comment:
With the upcoming release of Dulwich you can also use repo.object_store.iter_prefix, which avoids scanning the entire repository.

Reply:
Shall we wait till the next release and use the new method?

Reply:
I do not think we have to wait unless @jelmer says there will be a release in the next days / few weeks. (It will take at least some weeks until the next Poetry release and I think it makes sense to include this with or without the improvement.)

Reply:
FWIW There's a new release out with this API now.

Reply:
```suggestion
        for sha in repo.object_store.iter_prefix(sha_prefix):
```

Reply:
I can see that this is implemented in **dulwich** `0.22.5`, which drops support of Python 3.8.
It is a reasonable change, but this pull-request branch still requires compatibility with `3.8`. I'm not sure it must be changed as a part of this PR...

@radoering?

Reply:
We have dropped Python 3.8 on the main branch already, so that should not be an issue

Reply:
@mikhainin As Secrus said. Just rebase your branch on main. That is required before merging anyway.

Reply:
Hmm, this direct replace causes:
```

~/src/poetry/src/poetry/vcs/git/backend.py:342: in _find_object_by_sha_prefix
    for sha in repo.object_store.iter_prefix(sha_prefix):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <DiskObjectStore('/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of-mikhailgalanin/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects')>
prefix = b'895d5f'

    def iter_prefix(self, prefix):
        if len(prefix) < 2:
            yield from super().iter_prefix(prefix)
            return
        seen = set()
        dir = prefix[:2].decode()
        rest = prefix[2:].decode()
>       for name in os.listdir(os.path.join(self.path, dir)):
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects/89'

.../virtualenvs/poetry-GxhiNo8B-py3.11/lib/python3.11/site-packages/dulwich/object_store.py:1042: FileNotFoundError



```